src/Workspaces/Core/Portable/Execution/Serialization/Checksum.cs

```diff
@@ -73,8 +75,8 @@ public override int GetHashCode()
                     hash = Hash.Combine((int)_checkSum[i], hash);
                 }
 
-                _lazyHash = hash;
-            }
+                _lazyHash = new StrongBox<int>(hash);
+            };
```

Comment:
For example, SubstitutedNamedTypeSymbol has this as GetHashCode:

```
        public override int GetHashCode()
        {
            if (_hashCode == 0)
            {
                _hashCode = this.ComputeHashCode();
            }

            return _hashCode;
        }
```

This is safe because it's only a read/write of a 32bit value (which is atomic in .net).  And they ensure that ComputeHashCode never actually returns 0 itself.  So that prevents the case where the hash code keeps getting computed.  

The problem in your first example was the use of Nullable.  It's larger than 32bits, so the write of it won't be atomic and you may see writes reordered.  Similar with StrongBox.  

If you switch to a normal int, and you ensure that computing the hash code never returns 0, then you'll be safe.